<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ingres">
<title>Using Ingres to obtain single-cell probabilistic Boolean networks from single-cell RNA-seq data • ingres</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using Ingres to obtain single-cell probabilistic Boolean networks from single-cell RNA-seq data">
<meta property="og:description" content="ingres">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ingres</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/ingres.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/CBigOxf/ingres/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="ingres_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Using Ingres to obtain single-cell probabilistic Boolean networks from single-cell RNA-seq data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/CBigOxf/ingres/blob/HEAD/vignettes/ingres.Rmd" class="external-link"><code>vignettes/ingres.Rmd</code></a></small>
      <div class="d-none name"><code>ingres.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview-of-ingres">Overview of Ingres<a class="anchor" aria-label="anchor" href="#overview-of-ingres"></a>
</h2>
<p>Gene regulatory networks (GRN) have long been used as a way to integrate heterogeneous data into a discrete model, and are very useful to generate actionable hypotheses on the mechanisms governing these biological processes. Here we present Ingres (Inferring Probabilistic Boolean Networks of Gene Regulation Using Protein Activity Enrichment Scores) an open-source tool that uses<br>
single-cell sequencing data and prior knowledge GRNs to produce a probabilistic<br>
Boolean network (PBN) per each cell and/or cluster of cells in the dataset. Ingres allows to better capture the differences between cell phenotypes, using a continuous measure of protein activity while still confined to the simplicity of a GRN. We believe Ingres will be useful to better understand the heterogeneous makeup of cell populations, to gain insight into the specific circuits that drive certain phenotypes, and to use expression and other omics to infer computational cellular models in bulk or single-cell data.</p>
<p>Specifically, we use a previously developed algorithm, VIPER,<span class="citation"><sup><a href="#ref-alvarezfunctional2016" role="doc-biblioref">1</a></sup></span> to infer protein activity starting from a gene expression matrix and a list of regulons —defined as the list of transcriptional targets of each protein. This computes a matrix of normalised enrichment scores (NES) for each protein. The expression of these regulons represent an optimal indirect method to measure the activity of a specific gene/protein (the nodes in the preexisting Boolean network). This constitutes a novel way to fit the PBN using single-cell RNAseq data.</p>
<p>For more details on the specific algorithms used, please refer to our publication describing Ingres<span class="citation"><sup><a href="#ref-victori_ingres_2022" role="doc-biblioref">2</a></sup></span>.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Victori, P. &amp; Buffa, F. M. Ingres: from single-cell RNA-seq data to single-cell probabilistic Boolean networks. 2022.09.04.506528 Preprint at <a href="https://doi.org/10.1101/2022.09.04.506528" class="external-link uri">https://doi.org/10.1101/2022.09.04.506528</a> (2022).</p>
</div>
<div class="section level2">
<h2 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h2>
<p>Most of Ingres dependencies are available on CRAN. Some of them need to be installed from Bioconductor:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"viper"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"AnnotationDbi"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"aracne.networks"</span><span class="op">)</span> <span class="co"># optional dependency</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="creating-the-ingres-object">Creating the Ingres object<a class="anchor" aria-label="anchor" href="#creating-the-ingres-object"></a>
</h2>
<p>First we load the Ingres package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/CBigOxf/ingres" class="external-link">ingres</a></span><span class="op">)</span></span></code></pre></div>
<p>Then we need to create an Ingres object that will hold all the necessary data. Ingres contains example data to show how to use its functions, we will print them in this vignette so their structure can be seen. There are two main ways of creating an Ingres object. The easiest one is using a <code>Seurat</code><span class="citation"><sup><a href="#ref-Seurat" role="doc-biblioref">3</a></sup></span> object with the scaling and clustering pipelines already performed (see the corresponding <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">introductory vignette to Seurat</a>):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">small_blca_wang</span><span class="op">)</span></span>
<span><span class="va">small_blca_wang</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 16314 features across 90 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (16314 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  1 dimensional reduction calculated: tsne</span></span></code></pre></div>
<p>We also need a network object in <code>tidygraph</code> format:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"network"</span><span class="op">)</span></span>
<span><span class="va">network</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tbl_graph: 30 nodes and 110 edges</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; #</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; # A directed multigraph with 1 component</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; #</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; # Node Data: 30 × 3 (active)</span></span></span>
<span><span class="co">#&gt;   kind  rule                              id              </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> input <span style="color: #949494;">""</span>                                DNAdamage       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> input <span style="color: #949494;">""</span>                                GrowthInhibitors</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> fate  <span style="color: #949494;">"</span>CyclinE1 or CyclinA<span style="color: #949494;">"</span>             Proliferation   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> fate  <span style="color: #949494;">"</span>E2F1_2 or TP53<span style="color: #949494;">"</span>                  Apoptosis       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> fate  <span style="color: #949494;">"</span>p21CIP or RB1 or RBL2<span style="color: #949494;">"</span>           Growth_Arrest   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> gene  <span style="color: #949494;">"</span>SPRY and not FGFR3 and not GRB2<span style="color: #949494;">"</span> EGFR            </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 24 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; # Edge Data: 110 × 3</span></span></span>
<span><span class="co">#&gt;    from    to sign    </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     8    30 positive</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>    14     3 positive</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>    15     3 positive</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 107 more rows</span></span></span></code></pre></div>
<p>Ingres includes several helper functions that can transform GinSim or <code>graphml</code> files to the <code>tidygraph</code> format.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the graphml example file</span></span>
<span><span class="va">filename</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_network.graphml"</span>, package <span class="op">=</span> <span class="st">"ingres"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Convert to tidygraph format</span></span>
<span><span class="fu"><a href="../reference/graphmlAsTidy.html">graphmlAsTidy</a></span><span class="op">(</span><span class="va">filename</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tbl_graph: 30 nodes and 110 edges</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; #</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; # A directed multigraph with 1 component</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; #</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; # Node Data: 30 × 3 (active)</span></span></span>
<span><span class="co">#&gt;   kind  rule                              id              </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> input <span style="color: #949494;">""</span>                                DNAdamage       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> input <span style="color: #949494;">""</span>                                GrowthInhibitors</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> fate  <span style="color: #949494;">"</span>CyclinE1 or CyclinA<span style="color: #949494;">"</span>             Proliferation   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> fate  <span style="color: #949494;">"</span>E2F1_2 or TP53<span style="color: #949494;">"</span>                  Apoptosis       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> fate  <span style="color: #949494;">"</span>p21CIP or RB1 or RBL2<span style="color: #949494;">"</span>           Growth_Arrest   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> gene  <span style="color: #949494;">"</span>SPRY and not FGFR3 and not GRB2<span style="color: #949494;">"</span> EGFR            </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 24 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; # Edge Data: 110 × 3</span></span></span>
<span><span class="co">#&gt;    from    to sign    </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     8    30 positive</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>    14     3 positive</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>    15     3 positive</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 107 more rows</span></span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the GinSim (.zginml) example file</span></span>
<span><span class="va">filename</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_ginsim.zginml"</span>, package <span class="op">=</span> <span class="st">"ingres"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert to graphml and store it in a temporary file</span></span>
<span><span class="va">temp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">gml</span> <span class="op">=</span> <span class="fu"><a href="../reference/ginmlToGraphml.html">ginmlToGraphml</a></span><span class="op">(</span><span class="va">filename</span>, dest <span class="op">=</span> <span class="va">temp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gml</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "&lt;?xml version=\"1.0\" encoding=\"UTF-8\"?&gt;&lt;graphml xmlns=\"http://graphml.graphdrawing.org/xmlns\"&gt;"</span></span>
<span><span class="co">#&gt; [2] "&lt;key attr.name=\"kind\" attr.type=\"string\" for=\"node\" id=\"kind\"/&gt;"                            </span></span>
<span><span class="co">#&gt; [3] "&lt;key attr.name=\"rule\" attr.type=\"string\" for=\"node\" id=\"rule\"/&gt;"                            </span></span>
<span><span class="co">#&gt; [4] "&lt;key attr.name=\"sign\" attr.type=\"string\" for=\"edge\" id=\"sign\"/&gt;"                            </span></span>
<span><span class="co">#&gt; [5] "&lt;graph edgedefault=\"directed\"&gt;"                                                                   </span></span>
<span><span class="co">#&gt; [6] "&lt;node id=\"DNA_Damage\"&gt;"</span></span>
<span></span>
<span><span class="co"># Convert to tidygraph</span></span>
<span><span class="fu"><a href="../reference/graphmlAsTidy.html">graphmlAsTidy</a></span><span class="op">(</span><span class="va">temp</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tbl_graph: 26 nodes and 71 edges</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; #</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; # A directed simple graph with 1 component</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; #</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; # Node Data: 26 × 3 (active)</span></span></span>
<span><span class="co">#&gt;   kind  rule                                             id        </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> input <span style="color: #949494;">""</span>                                               DNA_Damage</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> gene  <span style="color: #949494;">"</span>DNA_Damage and (E2F1 or not Wip1 or not HDAC1)<span style="color: #949494;">"</span> ATM       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> gene  <span style="color: #949494;">"</span>ATM and not Wip1<span style="color: #949494;">"</span>                               p38MAPK   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> gene  <span style="color: #949494;">"</span>(not Wip1 or p53) and not ATM<span style="color: #949494;">"</span>                  Mdm2      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> gene  <span style="color: #949494;">"</span>ATM or (not Mdm2 and p38MAPK)<span style="color: #949494;">"</span>                  p53       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> gene  <span style="color: #949494;">"</span>p53_A<span style="color: #949494;">"</span>                                          Wip1      </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 20 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; # Edge Data: 71 × 3</span></span></span>
<span><span class="co">#&gt;    from    to sign    </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>    18     8 negative</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>    18     7 negative</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>    18    17 negative</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 68 more rows</span></span></span></code></pre></div>
<p>A data frame with a row per node in the network and its corresponding gene symbol, in case nodes in the network have different names:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"network_genes"</span><span class="op">)</span></span>
<span><span class="va">network_genes</span></span>
<span><span class="co">#&gt;        node   symbol</span></span>
<span><span class="co">#&gt; 1      EGFR     EGFR</span></span>
<span><span class="co">#&gt; 2     FGFR3    FGFR4</span></span>
<span><span class="co">#&gt; 3       RAS     HRAS</span></span>
<span><span class="co">#&gt; 4       RAS     KRAS</span></span>
<span><span class="co">#&gt; 5       RAS     NRAS</span></span>
<span><span class="co">#&gt; 6      E2F1     E2F1</span></span>
<span><span class="co">#&gt; 7  CyclinD1    CCND1</span></span>
<span><span class="co">#&gt; 8  CyclinE1    CCNE1</span></span>
<span><span class="co">#&gt; 9   CyclinA    CCNA1</span></span>
<span><span class="co">#&gt; 10  CyclinA    CCNA2</span></span>
<span><span class="co">#&gt; 11   CDC25A   CDC25A</span></span>
<span><span class="co">#&gt; 12 p16INK4a p16INK4a</span></span>
<span><span class="co">#&gt; 13      RB1      RB1</span></span>
<span><span class="co">#&gt; 14     RBL2     RBL2</span></span>
<span><span class="co">#&gt; 15   p21CIP   CDKN1A</span></span>
<span><span class="co">#&gt; 16      ATM      ATM</span></span>
<span><span class="co">#&gt; 17  CHEK1_2    CHEK1</span></span>
<span><span class="co">#&gt; 18  CHEK1_2    CHEK1</span></span>
<span><span class="co">#&gt; 19     MDM2     MDM2</span></span>
<span><span class="co">#&gt; 20     TP53     TP53</span></span>
<span><span class="co">#&gt; 21   p14ARF   CDKN2A</span></span>
<span><span class="co">#&gt; 22     PTEN     PTEN</span></span>
<span><span class="co">#&gt; 23     PI3K   PIK3CA</span></span>
<span><span class="co">#&gt; 24      AKT     AKT1</span></span>
<span><span class="co">#&gt; 25      AKT     AKT2</span></span>
<span><span class="co">#&gt; 26      AKT     AKT3</span></span>
<span><span class="co">#&gt; 27     GRB2     GRB2</span></span>
<span><span class="co">#&gt; 28     SPRY    SPRY2</span></span></code></pre></div>
<p>We can use the function <code>createNetworkGenesTemplate</code> to generate a network_genes file with the node names prepopulated. If using RStudio interactively, the function opens this file for us to modify as needed.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># store and modify = FALSE just for demonstration.</span></span>
<span><span class="fu"><a href="../reference/createNetworkGenesTemplate.html">createNetworkGenesTemplate</a></span><span class="op">(</span><span class="va">network</span>, store <span class="op">=</span> <span class="cn">FALSE</span>, modify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 25 × 2</span></span></span>
<span><span class="co">#&gt;    node     symbol  </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> EGFR     EGFR    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> FGFR3    FGFR3   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> RAS      RAS     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> E2F1_1   E2F1_1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> E2F1_2   E2F1_2  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> E2F3_1   E2F3_1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> E2F3_2   E2F3_2  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> CyclinD1 CyclinD1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> CyclinE1 CyclinE1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> CyclinA  CyclinA </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 15 more rows</span></span></span>
<span><span class="co"># Then, modify as needed.</span></span></code></pre></div>
<p>Then, we can create the Ingres object:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ing</span> <span class="op">=</span> <span class="fu"><a href="../reference/createIngresObjectFromSeurat.html">createIngresObjectFromSeurat</a></span><span class="op">(</span></span>
<span>  seurat.object <span class="op">=</span> <span class="va">small_blca_wang</span>,</span>
<span>  seurat.assay  <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  slot          <span class="op">=</span> <span class="st">"data"</span>,</span>
<span>  network.genes <span class="op">=</span> <span class="va">network_genes</span>,</span>
<span>  network       <span class="op">=</span> <span class="va">network</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ing</span></span>
<span><span class="co">#&gt; An ingres object</span></span>
<span><span class="co">#&gt; --------------------</span></span>
<span><span class="co">#&gt; 16314 features across 90 samples</span></span>
<span><span class="co">#&gt; 9 clusters</span></span>
<span><span class="co">#&gt; Its Boolean network has 30 nodes</span></span>
<span><span class="co">#&gt; Viper not performed</span></span></code></pre></div>
<p>An Ingres object can also be created without a <code>Seurat</code> object, by providing a expression matrix, with cells in rows and genes in columns, and a idents data.frame with each cell barcode and its corresponding cluster or subpopulation:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp</span> <span class="op">=</span> <span class="va">ing</span><span class="op">@</span><span class="va">expression</span></span>
<span><span class="va">exp</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="co">#&gt;              sample1@AGCTTGAGTTCCTCCA-1 sample1@CAGCAGCCAGTCTTCC-1</span></span>
<span><span class="co">#&gt; RP11-34P13.7                          0                          0</span></span>
<span><span class="co">#&gt; AL627309.1                            0                          0</span></span>
<span></span>
<span><span class="va">idents</span> <span class="op">=</span> <span class="va">ing</span><span class="op">@</span><span class="va">idents</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">idents</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         cell cluster</span></span>
<span><span class="co">#&gt; 1 sample1@AGCTTGAGTTCCTCCA-1       5</span></span>
<span><span class="co">#&gt; 2 sample1@CAGCAGCCAGTCTTCC-1       5</span></span>
<span><span class="co">#&gt; 3 sample1@CTAGCCTGTTAGTGGG-1       5</span></span>
<span><span class="co">#&gt; 4 sample1@GATGAGGGTCTGGTCG-1       5</span></span>
<span><span class="co">#&gt; 5 sample1@TGACTTTTCAAGAAGT-1       5</span></span>
<span><span class="co">#&gt; 6 sample1@TGCGTGGGTGTTGGGA-1       5</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/createIngresObject.html">createIngresObject</a></span><span class="op">(</span></span>
<span>  expression.matrix <span class="op">=</span> <span class="va">exp</span>, </span>
<span>  idents            <span class="op">=</span> <span class="va">idents</span>,   </span>
<span>  network.genes     <span class="op">=</span> <span class="va">network_genes</span>,</span>
<span>  network           <span class="op">=</span> <span class="va">network</span><span class="op">)</span></span>
<span><span class="co">#&gt; An ingres object</span></span>
<span><span class="co">#&gt; --------------------</span></span>
<span><span class="co">#&gt; 16314 features across 90 samples</span></span>
<span><span class="co">#&gt; 9 clusters</span></span>
<span><span class="co">#&gt; Its Boolean network has 30 nodes</span></span>
<span><span class="co">#&gt; Viper not performed</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-viper">Running VIPER<a class="anchor" aria-label="anchor" href="#running-viper"></a>
</h2>
<p>The next step is to run VIPER. For more information about the algorithm VIPER uses, please refer to <a href="https://bioconductor.org/packages/release/bioc/vignettes/viper/inst/doc/viper.pdf" class="external-link">its documentation</a>.</p>
<p>Ingres runs VIPER on the provided gene expression matrix, using the regulon supplied by the user. This regulon can come from the ARACNe algorithm<span class="citation"><sup><a href="#ref-bassoreverse2005" role="doc-biblioref">4</a>,<a href="#ref-lachmannaracneap2016" role="doc-biblioref">5</a></sup></span> or from other databases such as Dorothea.<span class="citation"><sup><a href="#ref-garcia-alonsobenchmark2019" role="doc-biblioref">6</a></sup></span> If several regulons are provided, the metaVIPER algorithm [14] will be run instead.<br>
This algorithm is designed to integrate multiple interactomes, not necessarily tissue-matched, as VIPER needed to work accurately. This is most valuable in single-cell RNA-seq data, due to its heterogeneity, inherent noisiness and low sequencing depth.<span class="citation"><sup><a href="#ref-ding2018" role="doc-biblioref">7</a></sup></span> Indeed, metaVIPER has been shown to reduce bias and batch effect, generating reproducible protein-activity signatures [14]. Therefore, we recommend this method for Ingres. Collections of tumour regulons are available at Bioconductor.<span class="citation"><sup><a href="#ref-giorgiaracne2017" role="doc-biblioref">8</a></sup></span> The package <code>aracne.networks</code>, a suggested dependency for <em>Ingres</em>, contains ARACNe-inferred networks from TCGA tumor datasets.</p>
<p>VIPER produces a matrix of Normalised Enrichment Scores (NES) per each gene and cell.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using a single regulon for speed</span></span>
<span><span class="va">ing</span> <span class="op">=</span> <span class="fu"><a href="../reference/performViper.html">performViper</a></span><span class="op">(</span><span class="va">ing</span>, <span class="fu">aracne.networks</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/aracne.networks/man/regulonblca.html" class="external-link">regulonblca</a></span><span class="op">)</span></span>
<span><span class="va">ing</span></span>
<span><span class="co">#&gt; An ingres object</span></span>
<span><span class="co">#&gt; --------------------</span></span>
<span><span class="co">#&gt; 16314 features across 90 samples</span></span>
<span><span class="co">#&gt; 9 clusters</span></span>
<span><span class="co">#&gt; Its Boolean network has 30 nodes</span></span>
<span><span class="co">#&gt; Viper performed</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="computing-the-pbns">Computing the PBNs<a class="anchor" aria-label="anchor" href="#computing-the-pbns"></a>
</h2>
<p>Now we can generate PBNs for every cell and/or every cluster in our <em>Ingres</em> object. This will generate a PBN that is a copy of the provided Boolean network but with a second rule in each gene node that is a fixed 1 or 0 (always active or always inactive). The probability of these rules being chosen will be proportional to the NES computed by VIPER. By default, this step will re-scale NES to (-1,1) for each gene in each cell. This is, the gene with the highest NES in a given cell will be re-scaled to 1 and the gene with the lowest NES in that same cell will be re-scaled to -1, and every other gene will be re-scaled relative to that range. Then, all genes with a re-scaled NES&gt;0 will have a fixed 1 rule added to the corresponding network node, with a probability equals to its re-scaled NES. All genes with a re-scaled NES&lt;0 will have a fixed 0 ruled added, with a probability equals to [re-scaled NES]. The original rule for each node will have a probability equals to 1-[fixed rule probability].</p>
<p>The range (-1,1) can be changed by the user, in case a less intense effect by the RNA-seq data is desired. For example, if the range is inputed as (-0.5, 0.5), the gene with the highest expression will have a 0.5 probability of activating its fixed rule, the same chance as its original rule.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ing</span> <span class="op">=</span> <span class="fu"><a href="../reference/computePbnByCluster.html">computePbnBySingleCell</a></span><span class="op">(</span><span class="va">ing</span><span class="op">)</span></span>
<span><span class="va">ing</span></span>
<span><span class="co">#&gt; An ingres object</span></span>
<span><span class="co">#&gt; --------------------</span></span>
<span><span class="co">#&gt; 16314 features across 90 samples</span></span>
<span><span class="co">#&gt; 9 clusters</span></span>
<span><span class="co">#&gt; Its Boolean network has 30 nodes</span></span>
<span><span class="co">#&gt; Viper performed </span></span>
<span><span class="co">#&gt; PBNs computed for each cell</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ing</span><span class="op">@</span><span class="va">single.cell.pbn</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         cell cluster         AKT        ATM      CHEK1_2</span></span>
<span><span class="co">#&gt; 1 sample1@AAGGAGCCAGCTTCGG-1       2  0.01817583 -0.7989694  0.099920471</span></span>
<span><span class="co">#&gt; 2 sample1@ACAGCTAAGATCCCGC-1       1 -0.07269741 -1.0000000  0.219362784</span></span>
<span><span class="co">#&gt; 3 sample1@ACGATGTAGTACATGA-1       3 -0.01603311 -0.9357372  0.513787703</span></span>
<span><span class="co">#&gt; 4 sample1@ACGGAGACAAATTGCC-1       2 -0.07899067 -0.7880775  0.055078000</span></span>
<span><span class="co">#&gt; 5 sample1@ACGGGCTCAAACCTAC-1       2 -0.04870924 -0.8023977 -0.002197616</span></span>
<span><span class="co">#&gt; 6 sample1@ACTGATGAGAGCTGGT-1       8  0.01102953 -0.6630933  0.153858311</span></span>
<span><span class="co">#&gt;     CyclinA    CyclinD1    CyclinE1        E2F1         EGFR       FGFR3</span></span>
<span><span class="co">#&gt; 1 1.0000000 0.005835620 -0.12590699  0.09456779 -0.002438333 -0.06437469</span></span>
<span><span class="co">#&gt; 2 0.2526539 0.100580602  0.03804537  0.14432723 -0.102653311 -0.08783756</span></span>
<span><span class="co">#&gt; 3 0.4896386 0.210391219 -0.04145495  1.00000000 -0.378800533  0.14514210</span></span>
<span><span class="co">#&gt; 4 1.0000000 0.060062911 -0.06046783 -0.06633384 -0.135412201 -0.03497538</span></span>
<span><span class="co">#&gt; 5 1.0000000 0.003268819  0.04396283  0.40529261  0.229916691 -0.24036193</span></span>
<span><span class="co">#&gt; 6 0.1336534 0.492846995 -0.26039627 -0.69295991 -0.509692556 -0.37852354</span></span>
<span><span class="co">#&gt;          GRB2         MDM2    p14ARF     p21CIP       PI3K        PTEN</span></span>
<span><span class="co">#&gt; 1  0.31113702 -0.033343239 0.3929228 0.38233341 -0.4071378  0.29342925</span></span>
<span><span class="co">#&gt; 2  0.32854332 -0.094203638 0.3691654 0.42156174 -0.7358486  0.35724316</span></span>
<span><span class="co">#&gt; 3 -0.07954781  0.054171975 0.5303340 0.01893101 -0.5368198 -0.05223321</span></span>
<span><span class="co">#&gt; 4  0.14676515 -0.033723994 0.4598230 0.29802852 -0.5897442  0.20687463</span></span>
<span><span class="co">#&gt; 5  0.05351330  0.008933491 0.3479199 0.37278246 -0.4302515  0.27558028</span></span>
<span><span class="co">#&gt; 6  0.37222843 -0.326267067 0.1103229 0.37679082 -1.0000000  0.05096307</span></span>
<span><span class="co">#&gt;         RAS         RB1       RBL2       SPRY        TP53</span></span>
<span><span class="co">#&gt; 1 0.3769146 -0.02103803 -0.2541966 -0.2188479 -0.10594942</span></span>
<span><span class="co">#&gt; 2 0.6074293 -0.15998474 -0.2668762 -0.2855518  0.10555768</span></span>
<span><span class="co">#&gt; 3 0.5145371 -0.38879500 -0.4278514 -0.3070425  0.15455175</span></span>
<span><span class="co">#&gt; 4 0.5643470 -0.18693366 -0.2616696 -0.1796798 -0.09579184</span></span>
<span><span class="co">#&gt; 5 0.5007909 -0.38687170 -0.3977755 -0.2124557  0.17860741</span></span>
<span><span class="co">#&gt; 6 0.4309240  0.66462958  0.3797944 -0.2995955 -0.12893251</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/computePbnByCluster.html">computePbnBySingleCell</a></span><span class="op">(</span><span class="va">ing</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span><span class="op">@</span><span class="va">single.cell.pbn</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         cell cluster          AKT        ATM      CHEK1_2</span></span>
<span><span class="co">#&gt; 1 sample1@AAGGAGCCAGCTTCGG-1       2  0.009087914 -0.3994847  0.049960235</span></span>
<span><span class="co">#&gt; 2 sample1@ACAGCTAAGATCCCGC-1       1 -0.036348706 -0.5000000  0.109681392</span></span>
<span><span class="co">#&gt; 3 sample1@ACGATGTAGTACATGA-1       3 -0.008016553 -0.4678686  0.256893851</span></span>
<span><span class="co">#&gt; 4 sample1@ACGGAGACAAATTGCC-1       2 -0.039495337 -0.3940388  0.027539000</span></span>
<span><span class="co">#&gt; 5 sample1@ACGGGCTCAAACCTAC-1       2 -0.024354619 -0.4011988 -0.001098808</span></span>
<span><span class="co">#&gt; 6 sample1@ACTGATGAGAGCTGGT-1       8  0.005514765 -0.3315466  0.076929156</span></span>
<span><span class="co">#&gt;     CyclinA    CyclinD1    CyclinE1        E2F1         EGFR       FGFR3</span></span>
<span><span class="co">#&gt; 1 0.5000000 0.002917810 -0.06295350  0.04728390 -0.001219166 -0.03218735</span></span>
<span><span class="co">#&gt; 2 0.1263270 0.050290301  0.01902269  0.07216362 -0.051326656 -0.04391878</span></span>
<span><span class="co">#&gt; 3 0.2448193 0.105195610 -0.02072748  0.50000000 -0.189400267  0.07257105</span></span>
<span><span class="co">#&gt; 4 0.5000000 0.030031455 -0.03023391 -0.03316692 -0.067706100 -0.01748769</span></span>
<span><span class="co">#&gt; 5 0.5000000 0.001634409  0.02198142  0.20264630  0.114958345 -0.12018097</span></span>
<span><span class="co">#&gt; 6 0.0668267 0.246423498 -0.13019813 -0.34647995 -0.254846278 -0.18926177</span></span>
<span><span class="co">#&gt;          GRB2         MDM2     p14ARF      p21CIP       PI3K        PTEN</span></span>
<span><span class="co">#&gt; 1  0.15556851 -0.016671619 0.19646142 0.191166703 -0.2035689  0.14671463</span></span>
<span><span class="co">#&gt; 2  0.16427166 -0.047101819 0.18458272 0.210780872 -0.3679243  0.17862158</span></span>
<span><span class="co">#&gt; 3 -0.03977391  0.027085987 0.26516699 0.009465505 -0.2684099 -0.02611660</span></span>
<span><span class="co">#&gt; 4  0.07338257 -0.016861997 0.22991148 0.149014262 -0.2948721  0.10343732</span></span>
<span><span class="co">#&gt; 5  0.02675665  0.004466745 0.17395994 0.186391231 -0.2151258  0.13779014</span></span>
<span><span class="co">#&gt; 6  0.18611422 -0.163133533 0.05516143 0.188395408 -0.5000000  0.02548154</span></span>
<span><span class="co">#&gt;         RAS         RB1       RBL2        SPRY        TP53</span></span>
<span><span class="co">#&gt; 1 0.1884573 -0.01051901 -0.1270983 -0.10942395 -0.05297471</span></span>
<span><span class="co">#&gt; 2 0.3037147 -0.07999237 -0.1334381 -0.14277588  0.05277884</span></span>
<span><span class="co">#&gt; 3 0.2572686 -0.19439750 -0.2139257 -0.15352125  0.07727588</span></span>
<span><span class="co">#&gt; 4 0.2821735 -0.09346683 -0.1308348 -0.08983989 -0.04789592</span></span>
<span><span class="co">#&gt; 5 0.2503954 -0.19343585 -0.1988878 -0.10622785  0.08930371</span></span>
<span><span class="co">#&gt; 6 0.2154620  0.33231479  0.1898972 -0.14979777 -0.06446626</span></span></code></pre></div>
<p>We can also compute a single PBN per each cluster. In that case, the median NES for each gene across all cells in that cluster is used to produce the PBN:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ing</span> <span class="op">=</span> <span class="fu"><a href="../reference/computePbnByCluster.html">computePbnByCluster</a></span><span class="op">(</span><span class="va">ing</span><span class="op">)</span></span>
<span><span class="va">ing</span></span>
<span><span class="co">#&gt; An ingres object</span></span>
<span><span class="co">#&gt; --------------------</span></span>
<span><span class="co">#&gt; 16314 features across 90 samples</span></span>
<span><span class="co">#&gt; 9 clusters</span></span>
<span><span class="co">#&gt; Its Boolean network has 30 nodes</span></span>
<span><span class="co">#&gt; Viper performed </span></span>
<span><span class="co">#&gt; PBNs computed for each cell</span></span>
<span><span class="co">#&gt; PBNs computed for each cluster</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization-of-results">Visualization of results<a class="anchor" aria-label="anchor" href="#visualization-of-results"></a>
</h2>
<p>We can export any PBN for further analysis by providing its cell barcode or cluster id:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/produceNetworkForCell.html">produceNetworkForCell</a></span><span class="op">(</span><span class="va">ing</span>, <span class="st">"sample1@ACAGCTAAGATCCCGC-1"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/produceNetworkForCluster.html">produceNetworkForCluster</a></span><span class="op">(</span><span class="va">ing</span>, <span class="st">"1"</span><span class="op">)</span></span></code></pre></div>
<p>In the same way, we can plot any network:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/cellPbnPlot.html">cellPbnPlot</a></span><span class="op">(</span><span class="va">ing</span>, <span class="st">"sample1@ACAGCTAAGATCCCGC-1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ingres_files/figure-html/network-plotting-1.png" width="100%"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/clusterPbnPlot.html">clusterPbnPlot</a></span><span class="op">(</span><span class="va">ing</span>, <span class="st">"1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ingres_files/figure-html/network-plotting-2.png" width="100%"></p>
<p>An overview of the results can be plotted as heatmaps:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/cellGenesHeatmap.html">cellGenesHeatmap</a></span><span class="op">(</span><span class="va">ing</span><span class="op">)</span></span></code></pre></div>
<p><img src="ingres_files/figure-html/heatmaps-1.png" width="100%"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/clusterGenesHeatmap.html">clusterGenesHeatmap</a></span><span class="op">(</span><span class="va">ing</span><span class="op">)</span></span></code></pre></div>
<p><img src="ingres_files/figure-html/heatmaps-2.png" width="100%"></p>
<p>Finally, any network can be exported as a <code>BoolNet</code><span class="citation"><sup><a href="#ref-BoolNet" role="doc-biblioref">9</a></sup></span> network. This package includes a variety of functions to analyse networks:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ing_network</span> <span class="op">=</span> <span class="fu"><a href="../reference/produceNetworkForCluster.html">produceNetworkForCluster</a></span><span class="op">(</span><span class="va">ing</span>, <span class="st">"1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">boolnet_network</span> <span class="op">=</span> <span class="fu"><a href="../reference/produceBoolnetNetwork.html">produceBoolnetNetwork</a></span><span class="op">(</span><span class="va">ing_network</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate a initial state where all nodes have state 1</span></span>
<span><span class="va">initial_state</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">network</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute one state transition from the given initial state.</span></span>
<span><span class="va">transition1</span> <span class="op">=</span> <span class="fu">BoolNet</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BoolNet/man/stateTransition.html" class="external-link">stateTransition</a></span><span class="op">(</span></span>
<span>  <span class="va">boolnet_network</span>, </span>
<span>  type  <span class="op">=</span> <span class="st">"probabilistic"</span>,</span>
<span>  state <span class="op">=</span> <span class="va">initial_state</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">transition1</span><span class="op">)</span></span>
<span><span class="co">#&gt;        DNAdamage GrowthInhibitors    Proliferation        Apoptosis </span></span>
<span><span class="co">#&gt;                1                1                1                1 </span></span>
<span><span class="co">#&gt;    Growth_Arrest             EGFR </span></span>
<span><span class="co">#&gt;                1                0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 20.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ingres_1.0.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] utf8_1.2.2             reticulate_1.26        tidyselect_1.2.0      </span></span>
<span><span class="co">#&gt;   [4] RSQLite_2.2.18         AnnotationDbi_1.58.0   htmlwidgets_1.5.4     </span></span>
<span><span class="co">#&gt;   [7] grid_4.2.1             Rtsne_0.16             munsell_0.5.0         </span></span>
<span><span class="co">#&gt;  [10] codetools_0.2-18       ragg_1.2.3             ica_1.0-3             </span></span>
<span><span class="co">#&gt;  [13] future_1.28.0          miniUI_0.1.1.1         withr_2.5.0           </span></span>
<span><span class="co">#&gt;  [16] spatstat.random_2.2-0  colorspace_2.0-3       progressr_0.11.0      </span></span>
<span><span class="co">#&gt;  [19] Biobase_2.56.0         highr_0.9              knitr_1.40            </span></span>
<span><span class="co">#&gt;  [22] Seurat_4.2.0           stats4_4.2.1           ROCR_1.0-11           </span></span>
<span><span class="co">#&gt;  [25] ggsignif_0.6.4         tensor_1.5             listenv_0.8.0         </span></span>
<span><span class="co">#&gt;  [28] labeling_0.4.2         GenomeInfoDbData_1.2.8 polyclip_1.10-0       </span></span>
<span><span class="co">#&gt;  [31] bit64_4.0.5            farver_2.1.1           rprojroot_2.0.3       </span></span>
<span><span class="co">#&gt;  [34] parallelly_1.32.1      vctrs_0.4.2            generics_0.1.3        </span></span>
<span><span class="co">#&gt;  [37] xfun_0.34              aracne.networks_1.22.0 R6_2.5.1              </span></span>
<span><span class="co">#&gt;  [40] GenomeInfoDb_1.32.4    graphlayouts_0.8.2     bitops_1.0-7          </span></span>
<span><span class="co">#&gt;  [43] spatstat.utils_2.3-1   cachem_1.0.6           promises_1.2.0.1      </span></span>
<span><span class="co">#&gt;  [46] scales_1.2.1           vroom_1.6.0            ggraph_2.1.0          </span></span>
<span><span class="co">#&gt;  [49] rgeos_0.5-9            gtable_0.3.1           globals_0.16.1        </span></span>
<span><span class="co">#&gt;  [52] goftest_1.2-3          tidygraph_1.2.2        rlang_1.0.6           </span></span>
<span><span class="co">#&gt;  [55] systemfonts_1.0.4      splines_4.2.1          rstatix_0.7.0         </span></span>
<span><span class="co">#&gt;  [58] lazyeval_0.2.2         broom_1.0.1            spatstat.geom_2.4-0   </span></span>
<span><span class="co">#&gt;  [61] yaml_2.3.6             reshape2_1.4.4         abind_1.4-5           </span></span>
<span><span class="co">#&gt;  [64] backports_1.4.1        httpuv_1.6.6           tools_4.2.1           </span></span>
<span><span class="co">#&gt;  [67] ggplot2_3.3.6          ellipsis_0.3.2         spatstat.core_2.4-4   </span></span>
<span><span class="co">#&gt;  [70] jquerylib_0.1.4        RColorBrewer_1.1-3     proxy_0.4-27          </span></span>
<span><span class="co">#&gt;  [73] BiocGenerics_0.42.0    ggridges_0.5.4         Rcpp_1.0.9            </span></span>
<span><span class="co">#&gt;  [76] plyr_1.8.7             zlibbioc_1.42.0        purrr_0.3.5           </span></span>
<span><span class="co">#&gt;  [79] RCurl_1.98-1.9         ggpubr_0.4.0           rpart_4.1.16          </span></span>
<span><span class="co">#&gt;  [82] deldir_1.0-6           pbapply_1.5-0          viridis_0.6.2         </span></span>
<span><span class="co">#&gt;  [85] cowplot_1.1.1          S4Vectors_0.34.0       zoo_1.8-11            </span></span>
<span><span class="co">#&gt;  [88] SeuratObject_4.1.2     ggrepel_0.9.1          cluster_2.1.4         </span></span>
<span><span class="co">#&gt;  [91] fs_1.5.2               magrittr_2.0.3         data.table_1.14.4     </span></span>
<span><span class="co">#&gt;  [94] scattermore_0.8        lmtest_0.9-40          RANN_2.6.1            </span></span>
<span><span class="co">#&gt;  [97] fitdistrplus_1.1-8     matrixStats_0.62.0     hms_1.1.2             </span></span>
<span><span class="co">#&gt; [100] patchwork_1.1.2        mime_0.12              evaluate_0.17         </span></span>
<span><span class="co">#&gt; [103] xtable_1.8-4           XML_3.99-0.11          IRanges_2.30.1        </span></span>
<span><span class="co">#&gt; [106] gridExtra_2.3          compiler_4.2.1         tibble_3.1.8          </span></span>
<span><span class="co">#&gt; [109] KernSmooth_2.23-20     crayon_1.5.2           htmltools_0.5.3       </span></span>
<span><span class="co">#&gt; [112] segmented_1.6-0        mgcv_1.8-40            later_1.3.0           </span></span>
<span><span class="co">#&gt; [115] tzdb_0.3.0             tidyr_1.2.1            DBI_1.1.3             </span></span>
<span><span class="co">#&gt; [118] tweenr_2.0.2           MASS_7.3-58.1          car_3.1-0             </span></span>
<span><span class="co">#&gt; [121] Matrix_1.5-1           readr_2.1.3            cli_3.4.1             </span></span>
<span><span class="co">#&gt; [124] parallel_4.2.1         igraph_1.3.5           pkgconfig_2.0.3       </span></span>
<span><span class="co">#&gt; [127] pkgdown_2.0.6          sp_1.5-0               plotly_4.10.0         </span></span>
<span><span class="co">#&gt; [130] spatstat.sparse_2.1-1  bslib_0.4.0            XVector_0.36.0        </span></span>
<span><span class="co">#&gt; [133] viper_1.30.0           stringr_1.4.1          digest_0.6.30         </span></span>
<span><span class="co">#&gt; [136] sctransform_0.3.5      RcppAnnoy_0.0.19       spatstat.data_2.2-0   </span></span>
<span><span class="co">#&gt; [139] Biostrings_2.64.1      BoolNet_2.1.7          rmarkdown_2.17        </span></span>
<span><span class="co">#&gt; [142] leiden_0.4.3           uwot_0.1.14            kernlab_0.9-31        </span></span>
<span><span class="co">#&gt; [145] shiny_1.7.2            lifecycle_1.0.3        nlme_3.1-160          </span></span>
<span><span class="co">#&gt; [148] jsonlite_1.8.2         carData_3.0-5          desc_1.4.2            </span></span>
<span><span class="co">#&gt; [151] viridisLite_0.4.1      fansi_1.0.3            pillar_1.8.1          </span></span>
<span><span class="co">#&gt; [154] lattice_0.20-45        KEGGREST_1.36.3        fastmap_1.1.0         </span></span>
<span><span class="co">#&gt; [157] httr_1.4.4             survival_3.4-0         glue_1.6.2            </span></span>
<span><span class="co">#&gt; [160] png_0.1-7              bit_4.0.4              ggforce_0.4.1         </span></span>
<span><span class="co">#&gt; [163] class_7.3-20           stringi_1.7.8          sass_0.4.2            </span></span>
<span><span class="co">#&gt; [166] mixtools_1.2.0         blob_1.2.3             textshaping_0.3.6     </span></span>
<span><span class="co">#&gt; [169] org.Hs.eg.db_3.15.0    memoise_2.0.1          dplyr_1.0.10          </span></span>
<span><span class="co">#&gt; [172] irlba_2.3.5.1          e1071_1.7-11           future.apply_1.9.1</span></span></code></pre></div>
</div>
<div class="section level2 unnumbered">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references">
<div id="ref-alvarezfunctional2016">
<p>1. Alvarez, M. J. <em>et al.</em> Functional characterization of somatic mutations in cancer using network-based inference of protein activity. <em>Nature Genetics</em> <strong>48</strong>, 838–847 (2016).</p>
</div>
<div id="ref-victori_ingres_2022">
<p>2. Victori, P. &amp; Buffa, F. M. Ingres: From single-cell RNA-seq data to single-cell probabilistic Boolean networks. 2022.09.04.506528 (2022) doi:<a href="https://doi.org/10.1101/2022.09.04.506528" class="external-link">10.1101/2022.09.04.506528</a>.</p>
</div>
<div id="ref-Seurat">
<p>3. Hao, Y. <em>et al.</em> Integrated analysis of multimodal single-cell data. (2021) doi:<a href="https://doi.org/10.1016/j.cell.2021.04.048" class="external-link">10.1016/j.cell.2021.04.048</a>.</p>
</div>
<div id="ref-bassoreverse2005">
<p>4. Basso, K. <em>et al.</em> Reverse engineering of regulatory networks in human b cells. <em>Nature Genetics</em> <strong>37</strong>, 382–390 (2005).</p>
</div>
<div id="ref-lachmannaracneap2016">
<p>5. Lachmann, A., Giorgi, F. M., Lopez, G. &amp; Califano, A. ARACNe-ap: Gene network reverse engineering through adaptive partitioning inference of mutual information. <em>Bioinformatics</em> <strong>32</strong>, 2233–2235 (2016).</p>
</div>
<div id="ref-garcia-alonsobenchmark2019">
<p>6. Garcia-Alonso, L., Holland, C. H., Ibrahim, M. M., Turei, D. &amp; Saez-Rodriguez, J. Benchmark and integration of resources for the estimation of human transcription factor activities. <em>Genome Research</em> <strong>29</strong>, 1363–1375 (2019).</p>
</div>
<div id="ref-ding2018">
<p>7. Ding, H. <em>et al.</em> Quantitative assessment of protein activity in orphan tissues and single cells using the metaVIPER algorithm. <em>Nature Communications</em> <strong>9</strong>, 1471 (2018).</p>
</div>
<div id="ref-giorgiaracne2017">
<p>8. Giorgi, F. M. <em>Aracne.networks: ARACNe-inferred gene networks from tcga tumor datasets</em>. (Bioconductor, 2017). doi:<a href="https://doi.org/10.18129/B9.BIOC.ARACNE.NETWORKS" class="external-link">10.18129/B9.BIOC.ARACNE.NETWORKS</a>.</p>
</div>
<div id="ref-BoolNet">
<p>9. Müssel, C., Hopfensitz, M. &amp; Kestler, H. A. BoolNet – an r package for generation, reconstruction and analysis of boolean networks. <em>Bioinformatics</em> <strong>26</strong>, 1378–1380 (2010).</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Pedro Victori, Francesca M. Buffa.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
